src/domain/safe/entities/transfer.entity.ts

```diff
@@ -1,4 +1,5 @@
 export interface Transfer {
+  type: string;
```

Comment:
Why was this added back? See https://github.com/5afe/safe-client-gateway-nest/pull/196

Reply:
Yes, I didn't notice that, probably was added back on a conflict resolution on my side. But, having that said, if we remove this field, we lose the transfer type coming from the Transaction Service, and how we could map the transfers without that information? I mean, each transfer type has specific rules/fields, and also the expected response includes a field with:
```javascript
"transferInfo": {
    "type": ...,
}
```

Maybe I'm missing something here 🤔

Reply:
The goal was that the `type` would only be used to specify to which object the payload would be serialised to. So AJV uses the `type` to check if the object should be checked against a `ERC20Transfer`, `ERC721Transfer` or `NativeTokenTransfer`.

Given that you then have this validation in place you wouldn't need a `type` on top of the different types available. We'd use something like type predicated to distinguish between the different types above: https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

Reply:
Maybe I'm missing something, but your proposal is to delete the type information from the transaction service response, and then recalculate it using another mechanism? 🤔

Reply:
> Maybe I'm missing something, but your proposal is to delete the type information from the transaction service response, and then recalculate it using another mechanism? 🤔

We are not deleting the type information. The type is in the interface itself. AJV uses the type to check which interface the payload should "serialised" to.

If you have:

```typescript
interface X {
  type: "X"
}

interface Y {
  type: "Y"
}

interface Z {
  type: "Z"
}
```

What would be the value of having the `type` as a `String` in addition to it being encapsulated in an interface that represents that same type?

Reply:
I tried to remove the `type: string;` line from the Transfer interface but I still not getting how you can implement something like the `getTransferByType` function ([this one](https://github.com/5afe/safe-client-gateway-nest/pull/206/files?diff=split&w=0#diff-e71537626c00f304d6f5a6e18f9d4fe64a39cc739a29deeeb36b8bd7df2693e4R54)) without having the type information. Do you propose using type predicates to infer the type of the transfer? I'm not sure how we can use the interfaces directly here (they are a TS construct witch is not available at runtime as you know) so the only way I see is to use type predicates to rebuild a piece of information which is already in the object 🤔 

Could you provide a concrete example to build the `Transfer` in `getTransferByType`?

Reply:
In `src/domain/safe/entities/transfer.entity.ts` you can add the following Type Predicates:

```typescript
export function isERC20Transfer(transfer: Transfer): transfer is ERC20Transfer {
  const erc20Transfer = transfer as ERC20Transfer;
  return (
    erc20Transfer.tokenAddress !== undefined &&
    erc20Transfer.value !== undefined
  );
}

export function isERC721Transfer(
  transfer: Transfer,
): transfer is ERC721Transfer {
  const erc721Transfer = transfer as ERC721Transfer;
  return (
    erc721Transfer.tokenAddress !== undefined &&
    erc721Transfer.tokenId !== undefined
  );
}

export function isNativeTokenTransfer(
  transfer: Transfer,
): transfer is NativeTokenTransfer {
  const nativeTokenTransfer = transfer as NativeTokenTransfer;
  return nativeTokenTransfer.value !== undefined;
}
```

And then `getTransferByType` can be written as follows:

```typescript
private async getTransferByType(
  chainId: string,
  domainTransfer: DomainTransfer,
): Promise<Transfer> {
  if (isERC20Transfer(domainTransfer)) {
    const { tokenAddress, value } = domainTransfer;
    const token = await this.getToken(chainId, tokenAddress);
    return new Erc20Transfer(
      token.address,
      value,
      token.name,
      token.symbol,
      token.logoUri,
      token.decimals,
    );
  } else if (isERC721Transfer(domainTransfer)) {
    const { tokenAddress, tokenId } = domainTransfer;
    const token = await this.getToken(chainId, tokenAddress);
    return new Erc721Transfer(
      token.address,
      tokenId,
      token.name,
      token.symbol,
      token.logoUri,
    );
  } else if (isNativeTokenTransfer(domainTransfer)) {
    return new NativeCoinTransfer(domainTransfer.value);
  } else {
    throw Error('Unknown transfer type');
  }
}
```

Reply:
Thanks for the example code 🙌🏻

I thought about that approach as well, but I saw some cons regarding it:
- We would be implementing/testing/maintaining that type predicates functions.
- The service would be doing extra computations to calculate something (the transfer type) that in fact, it's already present in the data.
- What happens if one of the conditions which define a type changes? (if `tokenId` is not `undefined` for the ERC721 in the future, for any reason) This would force us to implement the change in the transaction service and also here. It could lead to bugs for a change that happens outside of this service _domain_.

The obvious pro of using type predicates is we would have a cleaner/more elegant domain entity definition. What do you think?

Reply:
> Thanks for the example code 🙌🏻
> 
> I thought about that approach as well, but I saw some cons regarding it:
> 
> * We would be implementing/testing/maintaining that type predicates functions. (Using the provided `type` field comes for free.)

Correct. But I wouldn't say that `type` field coming from free is "good". We use the type to match with the schema but once we know the schema we shouldn't need the type. The `type` comes in the response to make it easier for the consumers of the API to know which schema to expect – once you know that, keeping the `type` is a code smell in my opinion.

When we design an interface that can have 3 interfaces implementing it (non-json related) we wouldn't add a type to each one or? 

> * The service would be doing extra computations to calculate something (the transfer type) that in fact, it's already present in the data.

This is assuming that checking the presence of some fields is computationally more expensive than doing a string comparison (to check the type) – I would have to check this.

> * What happens if one of the conditions which define a type changes? (if `tokenId` is not `undefined` for the ERC721 in the future, for any reason) This would force us to implement the change in the transaction service and also here. It could lead to bugs for a change that happens outside of this service _domain_.

This is contract breaking change. Every field in the domain should either have a value or be `null`. An `undefined` field means that it was removed from the payload. AJV would then have `null` in its place. If a nullable field becomes nullable we consider that it's an API breaking change.

If one of the conditions that define a type change wouldn't you also need to change AJV?

> The obvious pro of using type predicates is we would have a cleaner/more elegant domain entity definition. But I have some doubts about if it pays the price. What do you think?

I think you listed some good points but I'd say that for point 2) is not clear how it performs extra computations (even if it does are we optimising at this level) and for point 3) changing the API contract means that we need to check AJV. Also if the domain changes suddenly of course it can lead to bugs. But the domain is our contract with the backend.

Reply:
Ok, I get your point and I agree the domain should be clean. I tried to use the approach you proposed based on type predicates but I'm getting some mismatches between transfer types, because for some fields we are checking they are not `undefined` but in fact the transaction service is returning `null`. For example, this is a payload coming from the transaction service:

```
{
    "count": 2,
    "next": null,
    "previous": null,
    "results": [
        {
            "type": "ETHER_TRANSFER",
            "executionDate": "2022-08-04T12:44:22Z",
            "blockNumber": 7345420,
            "transactionHash": "0x5009500e71ff0dc12ddce3cbe64fd346df9bce9f2413f4dd4c4e82877313112b",
            "to": "0x4127839cdf4F73d9fC9a2C2861d8d1799e9DF40C",
            "value": "513839314563540",
            "tokenId": null,
            "tokenAddress": null,
            "tokenInfo": null,
            "from": "0x283Af0B28c62C092C9727F1Ee09c02CA627EB7F5"
        },
        {
            "type": "ERC721_TRANSFER",
            "executionDate": "2022-08-04T12:44:22Z",
            "blockNumber": 7345420,
            "transactionHash": "0x5009500e71ff0dc12ddce3cbe64fd346df9bce9f2413f4dd4c4e82877313112b",
            "to": "0x4127839cdf4F73d9fC9a2C2861d8d1799e9DF40C",
            "value": null,
            "tokenId": "110380171670080793190909999014722746036502744273864122623133899039475607835745",
            "tokenAddress": "0x57f1887a8BF19b14fC0dF6Fd9B2acc9Af147eA85",
            "tokenInfo": {
                "type": "ERC721",
                "address": "0x57f1887a8BF19b14fC0dF6Fd9B2acc9Af147eA85",
                "name": "Ethereum Name Service",
                "symbol": "ENS",
                "decimals": null,
                "logoUri": "https://safe-transaction-assets.safe.global/tokens/logos/ENS.png"
            },
            "from": "0x283Af0B28c62C092C9727F1Ee09c02CA627EB7F5"
        }
    ]
}
```

So to solve this, we should use something like:

```javascript
export function isERC20Transfer(transfer: Transfer): transfer is ERC20Transfer {
  const erc20Transfer = transfer as ERC20Transfer;
  return (
    erc20Transfer.tokenAddress !== undefined &&
    erc20Transfer.tokenAddress !== null &&
    erc20Transfer.value !== undefined &&
    erc20Transfer.value !== null
  );
}

export function isERC721Transfer(
  transfer: Transfer,
): transfer is ERC721Transfer {
  const erc721Transfer = transfer as ERC721Transfer;
  return (
    erc721Transfer.tokenAddress !== undefined &&
    erc721Transfer.tokenAddress !== null &&
    erc721Transfer.tokenId !== undefined &&
    erc721Transfer.tokenId !== null
  );
}

export function isNativeTokenTransfer(
  transfer: Transfer,
): transfer is NativeTokenTransfer {
  const nativeTokenTransfer = transfer as NativeTokenTransfer;
  return (
    nativeTokenTransfer.value !== undefined &&
    nativeTokenTransfer.value !== null
  );
}
```

What do you think about using nulls here? Not sure if we are adding too much complexity here 🤔 

Apart from that, maybe you already notice, but these type predicates are not mutually exclusive. I mean, a transfer with non-null `tokenAddress` and non-null `value` satisfies both `isERC20Transfer` and `isNativeTokenTransfer`, so we have to keep this in mind when building the `if-else` statements in `getTransferByType` or in other places: the order of the conditions matters. We wouldn't have this complexity if we just rely on the `type` field coming from the source.

I don't have a strong opinion here, just checking we are considering all the implications of this 🙂

Reply:
As I stated before I believe that having a type in a domain entity is strange. You'd need to explicitly cast depending on the type and then access the respective properties.

Type predicates do also perform a cast (implicit cast instead of an explicit one) but they work with the structure of the object to determine the type and not with a string value that defines it in a JSON payload.

If you believe that having type predicates adds complexity in this case and you'd see that having a type as a string is better then I'd say to go for it. I'll leave this decision to you.

Reply:
I do see value in using type predicates, having implicit casting, and having a clearer domain. I think the predicates are a great tool. I was just pointing out some concerns about their use in this particular use case. But I think we can go that way and see how it evolves, I took your suggested approach in [697dda9](https://github.com/5afe/safe-client-gateway-nest/pull/206/commits/697dda95b51318e373bcf339345ead59abb5e27d). Sorry for the long discussion, I just wanted to align on some cons it could have. Thanks for reviewing! 🙂

Reply:
I've addressed some of the concerns but it was my understanding that you see that it'd be more complex to maintain and write the type predicates. Therefore I'd approve whatever you decide and think it's easier to maintain.

If the having the type as a property is easier for maintaining and writing business logic around it then I'd say to go for it!